File: backend/mcp/resources.py
"""Resource registration and URI routing for FH6-HorizonTuner MCP."""

from __future__ import annotations

import json
from typing import Any
from urllib.parse import urlparse

from .protocol import McpError
from .service import HorizonTunerMcpService


class McpResourceManager:
    """Manages MCP Resources and maps URIs to service methods."""

    def __init__(self, service: HorizonTunerMcpService):
        self.service = service

    async def list_resources(self) -> list[dict[str, Any]]:
        """List statically discoverable resources and templates."""
        resources = [
            {
                "uri": "fh6://telemetry/live",
                "name": "Live Telemetry Snapshot",
                "description": "Latest real-time UDP telemetry frame and ingestion status.",
                "mimeType": "application/json",
            },
            {
                "uri": "fh6://settings/current",
                "name": "System & HUD Settings",
                "description": "Current user preferences, units, and HUD configuration.",
                "mimeType": "application/json",
            },
        ]

        # Add recent race sessions
        sessions = self.service.list_race_sessions(limit=5)
        for s in sessions:
            s_id = s.get("session_id")
            resources.append(
                {
                    "uri": f"fh6://telemetry/session/{s_id}",
                    "name": f"Race Session {s_id} ({s.get('car_name', 'Unknown')})",
                    "description": f"Session summary with {s.get('total_laps', 0)} laps.",
                    "mimeType": "application/json",
                }
            )

        # Add recent tuning captures
        captures = self.service.list_tuning_captures()[:5]
        for c in captures:
            c_id = c.get("capture_id")
            resources.append(
                {
                    "uri": f"fh6://capture/{c_id}",
                    "name": f"Tuning Capture {c_id}",
                    "description": f"Capture dataset ({c.get('samples_count', 0)} samples).",
                    "mimeType": "application/json",
                }
            )

        return resources

    async def read_resource(self, uri: str) -> dict[str, Any]:
        """Fetch and return JSON text content for a resource URI."""
        parsed = urlparse(uri)
        if parsed.scheme != "fh6":
            raise McpError(
                -32602, f"Unsupported URI scheme: {parsed.scheme}. Expected 'fh6://'"
            )

        path_parts = [p for p in parsed.path.strip("/").split("/") if p]
        host = parsed.netloc

        if host == "telemetry":
            if not path_parts or path_parts[0] == "live":
                data = self.service.get_live_telemetry_snapshot()
                return self._wrap_response(uri, data)
            if path_parts[0] == "session" and len(path_parts) > 1:
                session_id = path_parts[1]
                data = self.service.get_session_summary(session_id)
                if not data:
                    raise McpError(-32004, f"Session not found: {session_id}")
                return self._wrap_response(uri, data)

        elif host == "capture" and path_parts:
            capture_id = path_parts[0]
            data = self.service.get_capture_summary(capture_id)
            if not data:
                raise McpError(-32004, f"Capture dataset not found: {capture_id}")
            return self._wrap_response(uri, data)

        elif host == "car" and path_parts:
            car_id = path_parts[0]
            details = self.service.get_car_details(car_id)
            if not details:
                raise McpError(-32004, f"Car not found: {car_id}")
            caps = self.service.get_car_tuning_capabilities(car_id)
            return self._wrap_response(uri, {"details": details, "capabilities": caps})

        elif host == "tuning" and len(path_parts) >= 2:
            car_id = path_parts[0]
            save_name = path_parts[1]
            preset = self.service.get_tuning_preset(car_id, save_name)
            if not preset:
                raise McpError(-32004, f"Tuning preset not found: {car_id}/{save_name}")
            return self._wrap_response(uri, preset)

        elif host == "settings":
            settings = self.service.get_system_settings()
            hud = self.service.get_hud_configurations()
            return self._wrap_response(uri, {"settings": settings, "hud": hud})

        raise McpError(-32602, f"Unknown resource URI: {uri}")

    def _wrap_response(self, uri: str, data: Any) -> dict[str, Any]:
        return {
            "contents": [
                {
                    "uri": uri,
                    "mimeType": "application/json",
                    "text": json.dumps(data, indent=2, ensure_ascii=False),
                }
            ]
        }

File: backend-rust/src/mcp/resources.rs
use super::service::McpService;
use serde_json::{json, Value};

pub fn list(service: &McpService<'_>) -> Value {
    let mut out = vec![
        json!({"uri":"fh6://telemetry/live","name":"Live Telemetry Snapshot","description":"Latest real-time UDP telemetry frame and ingestion status.","mimeType":"application/json"}),
        json!({"uri":"fh6://settings/current","name":"System & HUD Settings","description":"Current user preferences, units, and HUD configuration.","mimeType":"application/json"}),
    ];
    for session in service.list_race_sessions(5, 0) {
        let id = session["session_id"].as_str().unwrap_or("");
        out.push(json!({"uri":format!("fh6://telemetry/session/{id}"),"name":format!("Race Session {id} ({})",session["car_name"].as_str().unwrap_or("Unknown")),"description":format!("Session summary with {} laps.",session["total_laps"].as_i64().unwrap_or(0)),"mimeType":"application/json"}));
    }
    for c in service.list_captures(None, None, None).into_iter().take(5) {
        let id = c["capture_id"].as_str().unwrap_or("");
        out.push(json!({"uri":format!("fh6://capture/{id}"),"name":format!("Tuning Capture {id}"),"description":format!("Capture dataset ({} samples).",c["samples_count"].as_i64().unwrap_or(0)),"mimeType":"application/json"}));
    }
    Value::Array(out)
}
fn wrap(uri: &str, data: Value) -> Value {
    json!({"contents":[{"uri":uri,"mimeType":"application/json","text":serde_json::to_string_pretty(&data).unwrap_or_else(|_|data.to_string())}]})
}
pub fn read(service: &McpService<'_>, uri: &str) -> Result<Value, (i64, String)> {
    let (scheme, rest) = uri.split_once("://").unwrap_or(("", uri));
    if scheme != "fh6" {
        return Err((
            -32602,
            format!("Unsupported URI scheme: {scheme}. Expected 'fh6://'"),
        ));
    }
    let mut p = rest.split('/');
    let host = p.next().unwrap_or("");
    let id = p.collect::<Vec<_>>().join("/");
    let data = match host {
        "telemetry" if id.is_empty() || id == "live" => service.get_live_telemetry_snapshot(),
        "telemetry" if id.starts_with("session/") => {
            let sid = id.trim_start_matches("session/");
            service
                .session_summary(sid)
                .ok_or_else(|| (-32004, format!("Session not found: {sid}")))?
        }
        "capture" => service
            .capture_summary(&id)
            .ok_or_else(|| (-32004, format!("Capture dataset not found: {id}")))?,
        "car" => {
            let d = service
                .car_details(&id)
                .ok_or_else(|| (-32004, format!("Car not found: {id}")))?;
            json!({"details":d,"capabilities":service.capabilities(&id,None)})
        }
        "tuning" if id.split('/').count() >= 2 => {
            let mut pieces = id.splitn(2, '/');
            let car = pieces.next().unwrap_or("");
            let save = pieces.next().unwrap_or("");
            service
                .preset(car, save)
                .ok_or_else(|| (-32004, format!("Tuning preset not found: {car}/{save}")))?
        }
        "settings" => json!({"settings":service.settings(),"hud":service.hud()}),
        _ => return Err((-32602, format!("Unknown resource URI: {uri}"))),
    };
    Ok(wrap(uri, data))
}
